cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -680,6 +816,7 @@ class InputState {
   Result<bool> AdvanceAndMemoize(OnType ts) {
     // Advance the right side row index until we reach the latest right row (for each key)
     // for the given left timestamp.
+    DEBUG_SYNC(node_, "Advancing input ", index_, DEBUG_MANIP(std::endl));
```

Comment:
`DEBUG_MANIP` is ok.  However, I think ending a debug statement with an endline is going to be more common than not ending with an endline (in fact, if we don't end with an endline, I don't think we guarantee things are flushed, and it would still be possible to get interleaved / out of order output).  It would be nice though if there was a `DEBUG_LINE` or `DEBUG_SYNC_LINE` or something that acted exactly like `DEBUG_SYNC` except it included the endline for us.

Reply:
I agree.

Reply:
Note that there are a [couple](https://github.com/apache/arrow/pull/34392/commits/85328512f4899ca92ce5404f442036f54d6ab3ee#diff-5493b6ae7ea2a4d5cfb581034c076e9c4be7608382168de6d1301ef67b6c01eeR1331) of [examples](https://github.com/apache/arrow/pull/34392/commits/85328512f4899ca92ce5404f442036f54d6ab3ee#diff-5493b6ae7ea2a4d5cfb581034c076e9c4be7608382168de6d1301ef67b6c01eeR1681) where `std::endl` is used in the middle of a message, so `DEBUG_MANIP` would be useful even if we had `DEBUG_SYNC_LINE`.